File: slcm/admission/page/admission_flow_dashboard/admission_flow_dashboard.py
import frappe
import json
from frappe import _
from frappe.utils import flt, getdate, add_days, today

@frappe.whitelist()
def get_dashboard_data(filters=None):
    if isinstance(filters, str):
        filters = json.loads(filters)
    
    if not filters:
        filters = {}

    query_filters = []
    if filters.get("admission_year"):
        query_filters.append(f"admission_year = {frappe.db.escape(filters.get('admission_year'))}")
    if filters.get("admission_cycle"):
        query_filters.append(f"admission_cycle = {frappe.db.escape(filters.get('admission_cycle'))}")
    if filters.get("campus"):
        query_filters.append(f"campus = {frappe.db.escape(filters.get('campus'))}")
    if filters.get("program"):
        query_filters.append(f"program = {frappe.db.escape(filters.get('program'))}")
    if filters.get("reservation_category"):
        query_filters.append(f"reservation_category = {frappe.db.escape(filters.get('reservation_category'))}")
    
    if filters.get("date_range"):
        date_range = filters.get("date_range")
        if isinstance(date_range, str):
             date_range = json.loads(date_range)
        if len(date_range) == 2:
            query_filters.append(f"creation >= {frappe.db.escape(date_range[0])}")
            query_filters.append(f"creation <= {frappe.db.escape(date_range[1] + ' 23:59:59')}")

    where_clause = " WHERE docstatus < 2 " # Exclude cancelled
    if query_filters:
        where_clause += " AND " + " AND ".join(query_filters)

    # 1. Summary Cards
    summary = get_summary_cards(where_clause)

    # 2. Funnel Data
    funnel = get_funnel_data(where_clause)

    # 3. Distribution by Campus
    campus_dist = get_distribution(where_clause, "campus")

    # 4. Distribution by Program
    program_dist = get_distribution(where_clause, "program")

    # 5. Application Trend (Last 30 days)
    trend = get_application_trend(where_clause)
    
    # 6. Category wise
    category_dist = get_distribution(where_clause, "reservation_category")

    # 7. Gender Distribution
    gender_dist = get_distribution(where_clause, "gender")

    # 8. Geographic Distribution (State)
    state_dist = get_distribution(where_clause, "state")

    # 9. Offer Status Breakdown
    offer_breakdown = get_offer_status_breakdown(where_clause)

    # 10. Fee Payment Status
    fee_payment_dist = get_fee_payment_status(where_clause)

    # 11. Yield Metrics
    yield_metrics = get_yield_metrics(summary)

    return {
        "summary": summary,
        "funnel": funnel,
        "campus_dist": campus_dist,
        "program_dist": program_dist,
        "trend": trend,
        "category_dist": category_dist,
        "gender_dist": gender_dist,
        "state_dist": state_dist,
        "offer_breakdown": offer_breakdown,
        "fee_payment_dist": fee_payment_dist,
        "yield_metrics": yield_metrics
    }

def get_summary_cards(where_clause):
    sql = f"""
        SELECT 
            COUNT(*) as total,
            SUM(CASE WHEN application_status NOT IN ('Draft') THEN 1 ELSE 0 END) as submitted,
            SUM(CASE WHEN application_status IN ('Selected', 'Offer Issued', 'Offer Accepted', 'Fee Paid', 'Accepted') THEN 1 ELSE 0 END) as selected,
            SUM(CASE WHEN application_status = 'Waitlisted' THEN 1 ELSE 0 END) as waitlisted,
            SUM(CASE WHEN application_status = 'Rejected' THEN 1 ELSE 0 END) as rejected,
            SUM(CASE WHEN application_status = 'Fee Paid' THEN 1 ELSE 0 END) as enrolled
        FROM `tabApplicant`
        {where_clause}
    """
    res = frappe.db.sql(sql, as_dict=1)[0]
    
    # Get Total Offers Issued (cumulative)
    sql_offers = f"""
        SELECT COUNT(*) FROM `tabApplicant` 
        {where_clause} AND application_status IN ('Offer Issued', 'Offer Accepted', 'Offer Declined', 'Offer Expired', 'Fee Paid')
    """
    offers_count = frappe.db.sql(sql_offers)[0][0]
    
    return {
        "total": res.total or 0,
        "submitted": res.submitted or 0,
        "selected": res.selected or 0,
        "waitlisted": res.waitlisted or 0,
        "rejected": res.rejected or 0,
        "enrolled": res.enrolled or 0,
        "offers": offers_count or 0
    }

def get_funnel_data(where_clause):
    # Stages for funnel
    stages = [
        {"label": "Submitted", "statuses": ["Submitted", "Selected", "Waitlisted", "Rejected", "Offer Accepted", "Offer Issued", "Offer Declined", "Offer Expired", "Fee Paid", "Accepted"]},
        {"label": "Selected", "statuses": ["Selected", "Offer Issued", "Offer Accepted", "Offer Declined", "Offer Expired", "Fee Paid", "Accepted"]},
        {"label": "Offer Issued", "statuses": ["Offer Issued", "Offer Accepted", "Offer Declined", "Offer Expired", "Fee Paid"]},
        {"label": "Offer Accepted", "statuses": ["Offer Accepted", "Fee Paid", "Accepted"]},
        {"label": "Fee Paid", "statuses": ["Fee Paid"]}
    ]
    
    data = []
    for stage in stages:
        status_list = ",".join([f"'{s}'" for s in stage["statuses"]])
        count_sql = f"SELECT COUNT(*) FROM `tabApplicant` {where_clause} AND application_status IN ({status_list})"
        count = frappe.db.sql(count_sql)[0][0]
        data.append({"label": stage["label"], "count": count})
    
    return data

def get_distribution(where_clause, field):
    sql = f"""
        SELECT IFNULL({field}, 'Not Specified') as label, COUNT(*) as count
        FROM `tabApplicant`
        {where_clause}
        GROUP BY {field}
        ORDER BY count DESC
        LIMIT 10
    """
    return frappe.db.sql(sql, as_dict=1)

def get_application_trend(where_clause):
    # Last 30 days trend
    date_limit = add_days(today(), -30)
    sql = f"""
        SELECT DATE(creation) as date, COUNT(*) as count
        FROM `tabApplicant`
        {where_clause} AND creation >= '{date_limit}'
        GROUP BY DATE(creation)
        ORDER BY date ASC
    """
    return frappe.db.sql(sql, as_dict=1)

def get_offer_status_breakdown(where_clause):
    sql = f"""
        SELECT 
            CASE 
                WHEN application_status = 'Offer Issued' THEN 'Pending Action'
                WHEN application_status IN ('Offer Accepted', 'Fee Paid', 'Accepted') THEN 'Accepted'
                WHEN application_status = 'Offer Declined' THEN 'Declined'
                WHEN application_status = 'Offer Expired' THEN 'Expired'
                ELSE 'Other'
            END as label,
            COUNT(*) as count
        FROM `tabApplicant`
        {where_clause} AND application_status IN ('Offer Issued', 'Offer Accepted', 'Offer Declined', 'Offer Expired', 'Fee Paid', 'Accepted')
        GROUP BY label
    """
    return frappe.db.sql(sql, as_dict=1)

def get_fee_payment_status(where_clause):
    # Join with Applicant to apply same filters (campus, cycle, year, etc.)
    sql = f"""
        SELECT 
            afa.status as label,
            COUNT(*) as count
        FROM `tabApplicant Fee Assignment` afa
        JOIN `tabApplicant` app ON afa.applicant = app.name
        {where_clause.replace('WHERE', 'WHERE afa.docstatus < 2 AND ')}
        GROUP BY label
        ORDER BY count DESC
    """
    # Note: where_clause already has 'WHERE docstatus < 2'. 
    # If where_clause has 'WHERE docstatus < 2 AND ...', replace 'WHERE' with 'WHERE afa.docstatus < 2 AND' 
    # is a bit tricky if it's already there.
    
    # Let's be safer with the replacement:
    final_where = where_clause.replace('admission_year', 'app.admission_year')\
                              .replace('admission_cycle', 'app.admission_cycle')\
                              .replace('campus', 'app.campus')\
                              .replace('program', 'app.program')\
                              .replace('reservation_category', 'app.reservation_category')\
                              .replace('gender', 'app.gender')\
                              .replace('state', 'app.state')\
                              .replace('creation', 'app.creation')\
                              .replace('docstatus', 'app.docstatus')

    sql = f"""
        SELECT 
            afa.status as label,
            COUNT(*) as count
        FROM `tabApplicant Fee Assignment` afa
        JOIN `tabApplicant` app ON afa.applicant = app.name
        {final_where}
        GROUP BY label
        ORDER BY count DESC
    """
    return frappe.db.sql(sql, as_dict=1)

def get_yield_metrics(summary):
    offers = summary.get("offers", 0)
    enrolled = summary.get("enrolled", 0)
    selected = summary.get("selected", 0)
    
    yield_rate = (enrolled / offers * 100) if offers > 0 else 0
    acceptance_rate = (enrolled / selected * 100) if selected > 0 else 0
    
    return {
        "yield_rate": round(yield_rate, 1),
        "acceptance_rate": round(acceptance_rate, 1)
    }

File: slcm/admission/page/admission_flow_dashboard/admission_flow_dashboard.js
frappe.pages['admission-flow-dashboard'].on_page_load = function (wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('Admission Flow Master Dashboard'),
        single_column: true
    });

    new AdmissionDashboard(wrapper);
}

class AdmissionDashboard {
    constructor(wrapper) {
        this.wrapper = $(wrapper);
        this.page = wrapper.page;
        this.filters = {};
        this.init();
    }

    init() {
        this.render_skeleton();
        this.setup_filters();
        this.refresh();
    }

    render_skeleton() {
        this.wrapper.html(`
            <div class="dashboard-wrapper">
                <div id="filter-container"></div>
                <div class="summary-grid" id="summary-cards"></div>
                
                <!-- Performance Metrics Row -->
                <div class="metrics-row mb-4">
                    <div id="yield-metrics-container" class="glass-card performance-summary p-4">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Conversion & Yield Performance')}</h6>
                        </div>
                        <div id="yield-metrics-grid" class="d-flex justify-content-around mt-4">
                            <!-- Populated dynamically -->
                        </div>
                    </div>
                </div>

                <div class="charts-grid">
                    <div id="funnel-chart-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Admission Pipeline')}</h6>
                            <span class="info-tag">${__('Funnel View')}</span>
                        </div>
                        <div id="funnel-chart"></div>
                    </div>
                    <div id="trend-chart-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Application Trends')}</h6>
                            <span class="info-tag">${__('Last 30 Days')}</span>
                        </div>
                        <div id="trend-chart"></div>
                    </div>

                    <!-- New Rows for Gender & Location -->
                    <div id="geographic-chart-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Geographic Distribution')}</h6>
                            <span class="info-tag">${__('by State')}</span>
                        </div>
                        <div id="geographic-chart"></div>
                    </div>

                    <div id="gender-chart-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Gender Diversity')}</h6>
                        </div>
                        <div id="gender-chart"></div>
                    </div>

                    <div id="campus-chart-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Distribution by Campus')}</h6>
                        </div>
                        <div id="campus-chart"></div>
                    </div>
                    <div id="program-chart-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Primary Program Choice')}</h6>
                        </div>
                        <div id="program-chart"></div>
                    </div>

                    <div id="fee-payment-card" class="chart-card glass-card">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Fee Payment Status')}</h6>
                        </div>
                        <div id="fee-payment-chart"></div>
                    </div>

                    <div id="offer-breakdown-card" class="chart-card glass-card full-width">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Offer Outcomes')}</h6>
                        </div>
                        <div id="offer-breakdown-chart"></div>
                    </div>

                    <div id="category-chart-card" class="chart-card glass-card full-width">
                        <div class="chart-header">
                            <h6 class="chart-title">${__('Category-wise Breakdown')}</h6>
                        </div>
                        <div id="category-chart"></div>
                    </div>
                </div>
            </div>
        `);
    }

    setup_filters() {
        let me = this;
        this.filter_bar = this.wrapper.find('#filter-container');
        this.filter_bar.addClass('filter-bar glass-card');

        // Add filters
        this.add_filter('admission_year', __('Admission Year'), 'Link', 'Admission Year');
        this.add_filter('admission_cycle', __('Admission Cycle'), 'Link', 'Admission Cycle');
        this.add_filter('campus', __('Campus'), 'Link', 'Campus');
        this.add_filter('program', __('Programme'), 'Link', 'Programme');
        this.add_filter('whether_scstobc_ncl', __('Category'), 'Link', 'Admission Category');
        this.add_filter('date_range', __('Date Range'), 'DateRange');

        // Clear button in the filter bar
        this.clear_btn = $(`<button class="btn btn-default btn-sm" style="margin-left: auto; height: 35px; align-self: flex-end; margin-bottom: 2px;">
            <i class="fa fa-refresh"></i> ${__('Reset')}
        </button>`).appendTo(this.filter_bar).on('click', () => {
            Object.keys(this.page.fields_dict).forEach(f => this.page.fields_dict[f].set_value(""));
            this.filters = {};
            this.refresh();
        });

        // Related Filter Logic: Admission Cycle depends on Admission Year
        this.page.fields_dict.admission_cycle.get_query = function () {
            return {
                filters: { 'admission_year': me.filters.admission_year || "" }
            };
        };

        // Initialize Defaults
        const year = frappe.defaults.get_user_default("admission_year");
        if (year) {
            this.page.fields_dict.admission_year.set_value(year);
            this.filters.admission_year = year;
        }

        const cycle = frappe.defaults.get_user_default("admission_cycle");
        if (cycle) {
            this.page.fields_dict.admission_cycle.set_value(cycle);
            this.filters.admission_cycle = cycle;
        }
    }

    add_filter(fieldname, label, fieldtype, options, reqd = 0) {
        let me = this;
        let field = this.page.add_field({
            fieldname: fieldname,
            label: label,
            fieldtype: fieldtype,
            options: options,
            reqd: reqd,
            change: function () {
                let val = this.get_value();
                me.filters[fieldname] = val;
                me.refresh();
            }
        });

        // Append to our custom filter container
        field.$wrapper.appendTo(this.filter_bar);

        // Ensure the label exists and is styled professionally
        let label_el = field.$wrapper.find('label');
        if (label_el.length === 0) {
            // Some Frappe versions don't add the label inside $wrapper for page fields
            $(`<label>${label}</label>`).prependTo(field.$wrapper);
            label_el = field.$wrapper.find('label');
        }

        // Apply robust styling for visibility
        field.$wrapper.css({
            'min-width': '220px',
            'margin-bottom': '10px !important',
            'display': 'flex',
            'flex-direction': 'column'
        });

        label_el.css({
            'display': 'block',
            'font-weight': '700',
            'font-size': '12px',
            'color': '#64748b',
            'margin-bottom': '8px',
            'text-transform': 'uppercase'
        });

        // Ensure input has height
        field.$wrapper.find('input, select').css({
            'height': '40px',
            'border-radius': '8px'
        });
    }

    refresh() {
        let me = this;

        this.wrapper.find('.charts-grid').css('opacity', '1').css('pointer-events', 'all');
        this.show_loader();

        frappe.call({
            method: 'slcm.admission.page.admission_flow_dashboard.admission_flow_dashboard.get_dashboard_data',
            args: { filters: this.filters },
            callback: function (r) {
                me.render_data(r.message);
            }
        });
    }

    show_loader() {
        // Maybe add some subtle opacity to cards
    }

    render_data(data) {
        if (!data) return;
        this.render_summary(data.summary);
        this.render_yield_metrics(data.yield_metrics || {});

        // Handle each chart with safety checks
        this.render_funnel_chart(data.funnel || []);
        this.render_trend_chart(data.trend || []);
        this.render_geographic_distribution(data.state_dist || []);
        this.render_gender_distribution(data.gender_dist || []);
        this.render_campus_distribution(data.campus_dist || []);
        this.render_program_distribution(data.program_dist || []);
        this.render_fee_payment_distribution(data.fee_payment_dist || []);
        this.render_offer_breakdown(data.offer_breakdown || []);
        this.render_category_distribution(data.category_dist || []);
    }

    render_yield_metrics(metrics) {
        let container = this.wrapper.find('#yield-metrics-grid').empty();
        const items = [
            { label: __('Final Yield Rate'), value: metrics.yield_rate || 0, sub: __('Enrolled / Total Offers'), color: 'bg-green' },
            { label: __('Applicant Conversion'), value: metrics.acceptance_rate || 0, sub: __('Enrolled / Selected'), color: 'bg-blue' }
        ];

        let html = items.map(i => `
            <div class="yield-item text-center">
                <div class="yield-label text-muted font-weight-bold mb-2">${i.label}</div>
                <div class="yield-value display-4 font-weight-bold mb-2" style="color: #1e293b">${i.value}%</div>
                <div class="progress mb-2" style="height: 10px; width: 250px; background: #f1f5f9; border-radius: 5px;">
                    <div class="progress-bar ${i.color}" role="progressbar" style="width: ${i.value}%; border-radius: 5px;"></div>
                </div>
                <div class="yield-sub text-muted small">${i.sub}</div>
            </div>
        `).join('<div class="vertical-divider mx-4" style="border-left: 1px solid #e2e8f0;"></div>');

        container.html(html);
    }

    render_summary(summary) {
        if (!summary) return;
        const cards = [
            { label: __('Total Applicants'), value: summary.total || 0, icon: 'fa fa-users', color: 'icon-blue' },
            { label: __('Submitted'), value: summary.submitted || 0, icon: 'fa fa-paper-plane', color: 'icon-purple' },
            { label: __('Selected'), value: summary.selected || 0, icon: 'fa fa-user-plus', color: 'icon-green' },
            { label: __('Offers'), value: summary.offers || 0, icon: 'fa fa-envelope-open', color: 'icon-yellow' },
            { label: __('Waitlisted'), value: summary.waitlisted || 0, icon: 'fa fa-clock-o', color: 'icon-teal' },
            { label: __('Enrolled'), value: summary.enrolled || 0, icon: 'fa fa-graduation-cap', color: 'icon-green' },
            { label: __('Rejected'), value: summary.rejected || 0, icon: 'fa fa-user-times', color: 'icon-red' },
        ];

        let html = cards.map(c => `
            <div class="summary-card glass-card" style="z-index: 1;">
                <div class="card-icon ${c.color}">
                    <i class="${c.icon}"></i>
                </div>
                <div>
                    <div class="card-label">${c.label}</div>
                    <div class="card-value">${c.value}</div>
                </div>
            </div>
        `).join('');

        this.wrapper.find('#summary-cards').html(html);
    }

    render_funnel_chart(funnel) {
        let container = $("#funnel-chart").empty();
        if (!funnel.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#funnel-chart", {
            data: {
                labels: funnel.map(f => f.label),
                datasets: [
                    {
                        name: __("Applicants"),
                        values: funnel.map(f => f.count)
                    }
                ]
            },
            type: 'bar', // Waterfall look with bar in descending order
            colors: ['#3b82f6'],
            height: 300,
            barOptions: {
                space_between_bars: 45
            }
        });
    }

    render_trend_chart(trend) {
        let container = $("#trend-chart").empty();
        if (!trend.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        // Line charts can crash with 1 data point in certain Frappe versions.
        // Fallback to bar in that case.
        let chart_type = trend.length > 1 ? 'line' : 'bar';

        new frappe.Chart("#trend-chart", {
            data: {
                labels: trend.map(t => {
                    if (!t.date) return "";
                    try {
                        let user_date = frappe.datetime.str_to_user(t.date);
                        return user_date ? user_date.split(',')[0] : t.date;
                    } catch (e) {
                        return t.date || "";
                    }
                }),
                datasets: [
                    {
                        name: __("Submissions"),
                        values: trend.map(t => parseInt(t.count || 0))
                    }
                ]
            },
            type: chart_type,
            height: 300,
            lineOptions: {
                hideDots: 0
            },
            colors: ['#8b5cf6']
        });
    }

    render_geographic_distribution(dist) {
        let container = $("#geographic-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#geographic-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [{ values: dist.map(d => d.count) }]
            },
            type: 'bar',
            height: 300,
            axisOptions: {
                xIsSeries: true
            },
            colors: ['#f43f5e']
        });
    }

    render_gender_distribution(dist) {
        let container = $("#gender-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#gender-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [{ values: dist.map(d => d.count) }]
            },
            type: 'pie',
            height: 300,
            colors: ['#ec4899', '#3b82f6', '#94a3b8']
        });
    }

    render_offer_breakdown(dist) {
        let container = $("#offer-breakdown-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#offer-breakdown-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [{ values: dist.map(d => d.count) }]
            },
            type: 'bar',
            height: 300,
            axisOptions: {
                xIsSeries: true
            },
            colors: ['#f59e0b']
        });
    }

    render_campus_distribution(dist) {
        let container = $("#campus-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#campus-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [{ values: dist.map(d => d.count) }]
            },
            type: 'bar', // Switched to bar for better readability when many campuses exist
            height: 300,
            axisOptions: {
                xIsSeries: true // Horizontal bar
            },
            colors: ['#10b981']
        });
    }

    render_program_distribution(dist) {
        let container = $("#program-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        // Use Horizontal Bar chart for programs. 
        // Pie charts fail when you have more than 5-7 categories as labels overlap.
        new frappe.Chart("#program-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [{ values: dist.map(d => d.count) }]
            },
            type: 'bar',
            height: 300,
            axisOptions: {
                xIsSeries: true // Makes it a horizontal bar chart
            },
            colors: ['#0ea5e9']
        });
    }

    render_category_distribution(dist) {
        let container = $("#category-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#category-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [
                    {
                        name: __("Applicants"),
                        values: dist.map(d => d.count)
                    }
                ]
            },
            type: 'bar',
            height: 300,
            axisOptions: {
                xIsSeries: true
            },
            colors: ['#f59e0b']
        });
    }

    render_fee_payment_distribution(dist) {
        let container = $("#fee-payment-chart").empty();
        if (!dist.length) {
            container.append(`<div class="text-center p-5 text-muted">${__('No data available')}</div>`);
            return;
        }

        new frappe.Chart("#fee-payment-chart", {
            data: {
                labels: dist.map(d => d.label),
                datasets: [{ values: dist.map(d => d.count) }]
            },
            type: 'pie',
            height: 300,
            colors: ['#10b981', '#f59e0b', '#3b82f6', '#f43f5e', '#64748b']
        });
    }
}
